### PR TITLE
fix(core): fix rate limit errors when deploying cloudwatch log groups 

### PR DESCRIPTION
**Pull request by xuetinp**
### Problem https://github.com/aws/aws-rfdk/issues/807
When the RFDK creates a CloudWatch Log group it does it via the CDK's [LogRetention ](https://docs.aws.amazon.com/cdk/api/v2/docs/aws-cdk-lib.aws_logs.LogRetention.html)construct. The CloudWatch Logs APIs that the LogRetention construct calls have a very low TPS limit (5 TPS). The result is that when deploying multiple RFDK constructs one can encounter rate limit errors in the deployment -- this is particularly seen, randomly, with the RFDK's integration tests.

### Solution 

CDK's LogRetention construct provides a way - [logRetentionRetryOptions](https://docs.aws.amazon.com/cdk/api/v2/docs/aws-cdk-lib.aws_logs.LogRetention.html#logretentionretryoptions) to configure the Javascript SDK that it's using to make the CloudWatch API calls. Using this, the RFDK should set higher max retries and retry delay. Currently default backoff retry base is 100ms and max retry count is 3, proposed value is 200ms and 7. This is a good combination as it achieves a lot of retries while the total function runtime is not too long. 

- Added retry option to all places where new LogRetention is created. 
- Extended log retention function timeout. Referencing https://github.com/aws/aws-cdk/blob/205e493e7bd6c5212f0ae374fdee28128ea49afe/packages/%40aws-cdk/aws-logs/lib/log-retention.ts#L122-L130

### Testing
* Created a CloudFormation app that iteratively initialize LogRetention constructs. Wrote a script to deploy constructs in parallel to reproduce the rate limit error. Verified proposed values for max retry count and retry delay prevent the error. 
* Added unit test to verify template is modified as expected.
* Build success.

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

File: packages/aws-rfdk/lib/core/lib/exporting-log-group.ts [hunk 1]
```diff
@@ -111,10 +113,25 @@ export class ExportingLogGroup extends Construct {
...
+    // as the default create log group TPS is only 5. Make sure to set the timeout of log retention function
+    // to be greater than total retry time. That's because if the function that is used for a custom resource
+    // doesn't exit properly, it'd end up in retries and may take cloud formation an hour to realize that
+    // the custom resource failed.
     const logRetention = new LogRetention(this, 'LogRetention', {
       logGroupName: props.logGroupName,
       retention: retentionInDays,
+      logRetentionRetryOptions: {
+        base: Duration.millis(200),
+        maxRetries: 7,
```
Comment (ddneilson) [inline packages/aws-rfdk/lib/core/lib/exporting-log-group.ts:126]: Checking the math... 

I think that the backoff code in the JS SDK is [here](https://github.com/aws/aws-sdk-js/blob/aa6049093165b84eac975aa50c3a44e23cbfe4af/lib/util.js#L866) ( called from [here](https://github.com/aws/aws-sdk-js/blob/aa6049093165b84eac975aa50c3a44e23cbfe4af/lib/service.js#L551) amongst other places ).

Looks like the total amount of delay time over all retries will be `(2^0 + 2^1 + ... + 2^(maxRetries-1} ) * 200ms`. With maxRetries of 6 that puts our max delay at 25.2s. So, we should be good with these settings and increasing the Lambda Function's timeout to 30s.

**Review by ddneilson — DISMISSED** [review]
This is perfect. Thank you!

File: packages/aws-rfdk/lib/core/lib/exporting-log-group.ts [hunk 2]
```diff
@@ -111,10 +113,25 @@ export class ExportingLogGroup extends Construct {
...
+      },
     });
+    const logRetentionFunctionLogicalId = 'LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8a';
```
Comment (jericht) [inline packages/aws-rfdk/lib/core/lib/exporting-log-group.ts]: Looks like we've grabbed this from the CDK code linked in the PR description. Could we also add a short comment here saying this is taken directly from the underlying CDK code as well?

File: packages/aws-rfdk/lib/core/lib/log-group-factory.ts [hunk 3]
```diff
@@ -64,7 +71,20 @@ export class LogGroupFactory {
         new LogRetention(scope, logWrapperId, {
           logGroupName: fullLogGroupName,
           retention,
+          logRetentionRetryOptions: {
+            base: Duration.millis(200),
+            maxRetries: 7,
+          },
         }).logGroupArn);
+
+    const logRetentionFunctionLogicalId = 'LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8a';
```
Comment (jericht) [inline packages/aws-rfdk/lib/core/lib/log-group-factory.ts]: Same thing here, let's add a short comment saying this is taken directly from the CDK code

File: packages/aws-rfdk/lib/core/lib/exporting-log-group.ts (hunk 2 above)
Comment (ddneilson) [reply packages/aws-rfdk/lib/core/lib/exporting-log-group.ts]: Good point. Breadcrumbs are useful...

Comment (xuetinp) [reply packages/aws-rfdk/lib/core/lib/exporting-log-group.ts]: Will do

File: packages/aws-rfdk/lib/core/lib/log-group-factory.ts (hunk 3 above)
Comment (xuetinp) [reply packages/aws-rfdk/lib/core/lib/log-group-factory.ts]: Will do

File: packages/aws-rfdk/lib/core/lib/exporting-log-group.ts [hunk 4]
```diff
@@ -111,10 +113,25 @@ export class ExportingLogGroup extends Construct {
...
+    const logRetentionFunction = Stack.of(this).node.tryFindChild(logRetentionFunctionLogicalId);
+    if (logRetentionFunction) {
+      const cfnFunction = logRetentionFunction.node.defaultChild as CfnFunction;
+      cfnFunction.addPropertyOverride('Timeout', 30);
+    }
```
Comment (jusiskin) [inline packages/aws-rfdk/lib/core/lib/exporting-log-group.ts]: Should we fail if we cannot find the log function? I'm wondering about the case where we upgrade CDK to a new version and the construct path changes.

This would be a breaking change in CDK, but we might not catch the integration break unless the developer doing the CDK upgrade knows to check here. Having this throw an exception would be a more explicit failure mode.

Same feedback applies to `LogGroupFactory` below.

What do you think?

File: packages/aws-rfdk/lib/core/lib/exporting-log-group.ts (hunk 2 above)
Comment (jusiskin) [inline packages/aws-rfdk/lib/core/lib/exporting-log-group.ts]: Strictly speaking, I don't believe this is a [logical ID](https://docs.aws.amazon.com/cdk/v2/guide/identifiers.html#identifiers_logical_ids). The [API docs](https://docs.aws.amazon.com/cdk/api/v2/docs/constructs.Node.html#trywbrfindwbrchildid) for `Node.tryFindChild(...)` document the argument with:

>   Identifier of direct child

And I believe this refers to the [construct ID](https://docs.aws.amazon.com/cdk/v2/guide/identifiers.html#identifiers_construct_ids) which is a unique ID within the scope of its parent in the [construct tree](https://docs.aws.amazon.com/cdk/v2/guide/constructs.html#constructs_tree).

TLDR: Can we rename this variable to `logRetentionFunctionConstructId`?

Same feedback applies to the similar change in the `LogGroupFactory` class below.

**Review by jusiskin — CHANGES_REQUESTED** [review]
Great work. I have two minor suggestions to improve this.

File: packages/aws-rfdk/lib/core/lib/exporting-log-group.ts (hunk 4 above)
Comment (xuetinp) [reply packages/aws-rfdk/lib/core/lib/exporting-log-group.ts]: Good callout! I will update.

File: packages/aws-rfdk/lib/core/lib/exporting-log-group.ts (hunk 2 above)
Comment (xuetinp) [reply packages/aws-rfdk/lib/core/lib/exporting-log-group.ts]: I referenced the variable name from https://github.com/aws/aws-cdk/blob/205e493e7bd6c5212f0ae374fdee28128ea49afe/packages/%40aws-cdk/aws-logs/lib/log-retention.ts#L122-L130. Actually I think this id may be both logical id and construct id. It's logical id because in my testing I created different stacks, but they all have the same logical id for the function in the output cloudformation template. Meanwhile, it's also construct id for LogRetentionFunction.

Comment (xuetinp) [reply packages/aws-rfdk/lib/core/lib/exporting-log-group.ts]: Updated it to `logRetentionFunctionConstructId` to reduce confusion.

File: packages/aws-rfdk/lib/core/lib/exporting-log-group.ts [hunk 5]
```diff
@@ -126,12 +126,11 @@ export class ExportingLogGroup extends Construct {
         maxRetries: 7,
       },
     });
-    const logRetentionFunctionLogicalId = 'LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8a';
-    const logRetentionFunction = Stack.of(this).node.tryFindChild(logRetentionFunctionLogicalId);
-    if (logRetentionFunction) {
-      const cfnFunction = logRetentionFunction.node.defaultChild as CfnFunction;
-      cfnFunction.addPropertyOverride('Timeout', 30);
-    }
+    // referenced from cdk code: https://github.com/aws/aws-cdk/blob/main/packages/@aws-cdk/aws-logs/lib/log-retention.ts#L123
```
Comment (jusiskin) [inline packages/aws-rfdk/lib/core/lib/exporting-log-group.ts]: nit: this URL is not pinned to a commit, so the line number may drift when CDK's default branch pointer is modified.

The best approach is to browse the git tag for the version RFDK currently targets (`2.33.0`) and link to that.

```suggestion
    // referenced from cdk code: https://github.com/aws/aws-cdk/blob/v2.33.0/packages/@aws-cdk/aws-logs/lib/log-retention.ts#L116
```

File: packages/aws-rfdk/lib/core/lib/exporting-log-group.ts [hunk 6]
```diff
@@ -111,10 +113,24 @@ export class ExportingLogGroup extends Construct {
...
       retention: retentionInDays,
+      logRetentionRetryOptions: {
+        base: Duration.millis(200),
+        maxRetries: 7,
+      },
     });
+    // referenced from cdk code: https://github.com/aws/aws-cdk/blob/main/packages/@aws-cdk/aws-logs/lib/log-retention.ts#L123
+    const logRetentionFunctionConstructId = 'LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8a';
+    const logRetentionFunction = Stack.of(this).node.tryFindChild(logRetentionFunctionConstructId)!;
+    const cfnFunction = logRetentionFunction.node.defaultChild as CfnFunction;
```
Comment (jusiskin) [inline packages/aws-rfdk/lib/core/lib/exporting-log-group.ts]: Just to confirm that this change was made in response to my previous suggestion where we'd throw an exception if we couldn't find the lambda function.

The belief here is that if the node without the specified construct ID does not exist, then `logRetentionFunction` will be `undefined` and line `132` will throw a:

```txt
Uncaught TypeError: Cannot read property 'node' of undefined
```

There may be an opportunity to make the error more developer-friendly since we have all the context here. How about this:

```suggestion
    const logRetentionFunction = Stack.of(this).node.tryFindChild(logRetentionFunctionConstructId);
    if (!logRetentionFunction) {
      throw new Error(`Could not find log retention Lambda Function with construct ID ${logRetentionFunctionConstructId} in construct tree`);
    }
    const cfnFunction = logRetentionFunction.node.defaultChild as CfnFunction;
```

Same feedback applies for `LogGroupFactory` below.

**Review by jusiskin — CHANGES_REQUESTED** [review]

File: packages/aws-rfdk/lib/core/lib/exporting-log-group.ts (hunk 5 above)
Comment (xuetinp) [reply packages/aws-rfdk/lib/core/lib/exporting-log-group.ts]: Will do

File: packages/aws-rfdk/lib/core/lib/exporting-log-group.ts (hunk 2 above)
Comment (jusiskin) [reply packages/aws-rfdk/lib/core/lib/exporting-log-group.ts]: Ah, I see their code uses similar naming. Oddly, in CDK's docs on [logical IDs](https://docs.aws.amazon.com/cdk/v2/guide/identifiers.html#identifiers_logical_ids):

>   Unique IDs serve as the logical identifiers

The [docs for unique IDs](https://docs.aws.amazon.com/cdk/v2/guide/identifiers.html#identifiers_unique_ids) say that a node's unique ID is derived from its construct path:

>   AWS CDK generates the necessary unique identifiers by concatenating the elements of the path and adding an 8-digit hash

The [path docs](https://docs.aws.amazon.com/cdk/v2/guide/identifiers.html#identifiers_paths) state:

>   We refer to the collection of IDs from a given construct, its parent construct, its grandparent, and so on to the root of the construct tree, as a path.

The root of the construct tree is the app. The stack construct ID should be a part of the path, so based on my understanding of the docs, you should get different logical IDs for different stacks within the same app. \*shrug\*

File: packages/aws-rfdk/lib/core/lib/exporting-log-group.ts (hunk 6 above)
Comment (xuetinp) [reply packages/aws-rfdk/lib/core/lib/exporting-log-group.ts]: The reason I didn't explicitly throw new error is that it will cause insufficient branch coverage and fail the build, and it's not easy to test that error branch as the log retention function is defined in cdk library. 

I tried using `findChild` function instead and it throws more explicit error. The code will be `const logRetentionFunction = Stack.of(this).node.findChild(logRetentionFunctionConstructId);` and error message will be like `Error: No child with id: 'LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8'`. I think this error message is a little better than undefined message above. What do you think if we switch to this?

Comment (jusiskin) [reply packages/aws-rfdk/lib/core/lib/exporting-log-group.ts]: Yeah, this combined with the stack trace will be sufficient context for the developer use-case I'd envisioned. Let's go with it.

File: packages/aws-rfdk/lib/core/lib/exporting-log-group.ts (hunk 2 above)
Comment (xuetinp) [reply packages/aws-rfdk/lib/core/lib/exporting-log-group.ts]: Yes, the LogRetention resource that I created in my TestStacks has a different logical ID per stack. The logical id is concatenation of construct ids of its parent construct, its grandparent etc. But the LogRetentionFunction resource has a fixed logical ID that is not related to any parent constructs. I'm not sure why, but it's interesting to know.

**Review by jericht — APPROVED** [review]
LGTM!

**Review by ddneilson — APPROVED** [review]

**Review by jusiskin — APPROVED** [review]
